automatic_push/rfc_push.py

```diff
@@ -0,0 +1,124 @@
+) -> PullRequestCreationResult:
+    if not new_files and not diff_files:
+        return PullRequestCreationResult(False, 'No files to update')
+    token = config.get('Secrets-Section', 'yang-catalog-token')
+    username = config.get('General-Section', 'repository-username')
+    try:
+        _update_files_locally(new_files, diff_files, forked_repo, config)
+        if not forked_repo.repo.index.diff(None):
+            return PullRequestCreationResult(False, 'No changed files found locally')
+        return _create_pull_request(forked_repo, username, token, logger, config)
+    except Exception as e:
+        logger.exception('Unexpected error occurred during an automatic PullRequest creation')
+        return PullRequestCreationResult(False, f'Unexpected error\n{e}')
+
+
+def _update_files_locally(new_files: list[str], diff_files: list[str], repo: ModifiableRepoUtil, config: ConfigParser):
+    ietf_directory = config.get('Directory-Section', 'ietf-directory')
+    rfc_directory = os.path.join(ietf_directory, 'YANG-rfc')
+    copy_to = os.path.join(repo.local_dir, 'standard/ietf/RFC')
+    cwd = os.getcwd()
+    os.chdir(copy_to)
+    for filename in new_files + diff_files:
+        file_path = os.path.join(rfc_directory, filename)
+        filename_without_revision = f'{filename.split("@")[0]}.yang'
+        if not os.path.exists(file_path):
+            continue
+        shutil.copy2(file_path, filename)
+        if filename not in new_files:
+            continue
+        if os.path.islink(filename_without_revision):
+            os.unlink(filename_without_revision)
+        os.symlink(filename, filename_without_revision)
+    os.chdir(cwd)
+
+
+def _create_pull_request(
+    forked_repo: ModifiableRepoUtil,
+    username: str,
+    token: str,
+    logger: logging.Logger,
+    config: ConfigParser,
+) -> PullRequestCreationResult:
+    is_prod = config.get('General-Section', 'is-prod') == 'True'
+    if not is_prod:
+        forked_repo.repo.git.reset('--hard')
+        return PullRequestCreationResult(
+            True,
+            'PullRequest creation is successful (switch to the PROD environment to actually create one)',
+        )
+    forked_repo.repo.git.add(all=True)
+    forked_repo.commit_all(message='Add new IETF RFC files')
+    forked_repo.repo.git.push('--set-upstream', 'origin', forked_repo.repo.active_branch)
+    open_pr_response = requests.get(
+        'https://api.github.com/repos/YangModels/yang/pulls',
+        headers={'Authorization': f'token {token}', 'Content-Type': 'application/vnd.github+json'},
+        data=json.dumps({'state': 'open', 'head': f'yang-catalog:{forked_repo.repo.active_branch}'}),
+    )
+    if prs := open_pr_response.json():
+        return PullRequestCreationResult(
+            False,
+            f'There is already an open PullRequest (files were updated in this PullRequest): {prs[0]["html_url"]}',
+        )
+    response = create_pull_request(
```

Comment:
what about passing these parameters as one dictionary object
I can see they consist mostly of strings and other dicts.
You can create some dataclass with name `PullRequestDetail` for example and pass it as 1 parameter.

Reply:
What would the advantage of this be here?

Reply:
Instead of passing 4 strings as parameters + one dict which contains headers, everything will be in one object.

Reply:
I created a dataclass for this purpose for the future (for example, ycJobs.py can use this function) but for now, I removed the manual creation of a PR because a Pr would anyway be created after the successful run of the GitHub Actions, so files will only be pushed to the repo

Reply:
> Instead of passing 4 strings as parameters + one dict which contains headers, everything will be in one object.

That's true, but the call to the constructor for the dataclass will still need to be called with all 5 of the parameters. I don't really see how that improves the situation. It would make sense if there were multiple places where we'd need the same set of arguments or if they all needed to be passed through several levels of functions, but that's not the case here.

Reply:
You are absolutely right, I didn't express myself properly. Instead of `dataclass`, I was thinking of `TypedDict`.

Reply:
I think it would be more beneficial to use a ```TypedDict``` only for the ```request_body``` param because it's not immediately understandable (moreover, when there are functions for creating, approving, merging, ...)
So, there will be ```TypedDicts``` for the request body of each of these functions, what do you think?